pkg/devfile/parser/reader.go

```diff
@@ -0,0 +1,107 @@
+package parser
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/devfile/library/pkg/testingutil/filesystem"
+	"github.com/devfile/library/pkg/util"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+	k8yaml "sigs.k8s.io/yaml"
+
+	routev1 "github.com/openshift/api/route/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// YamlSrc specifies the src of the yaml in either Path, URL or Data format
+type YamlSrc struct {
+	// Path is a relative or absolute yaml path.
+	Path string
+	// URL is the URL address of the specific yaml.
+	URL string
+	// Data is the yaml content in []byte format.
+	Data []byte
+}
+
+// ReadKubernetesYaml reads a yaml Kubernetes file from either the Path, URL or Data provided.
+// It returns Deployments, Services, Routes resources as the primary Kubernetes resources.
+// Other Kubernetes resources are returned as []byte type. Consumers interested in other Kubernetes resources
+// are expected to Unmarshal it to the struct of the respective resource.
+func ReadKubernetesYaml(src YamlSrc, fs filesystem.Filesystem) ([]appsv1.Deployment, []corev1.Service, []routev1.Route, [][]byte, error) {
+
+	var data []byte
+	var err error
+
+	if src.URL != "" {
+		data, err = util.DownloadFileInMemory(src.URL)
```

Comment:
Can we separate the logic to download / read the Kubernetes yaml file, from the logic to parse the files resources? I.e. maybe modify `ReadKubernetesYaml` to take in a byte array?

Reply:
updated